File: templates/default/controller/auth_controller.c
#include "../src/app.h"

#include <stdio.h>

static void respond_auth_error(int client, const char *message) {
  char escaped[512];
  char json[768];
  json_escape(escaped, sizeof(escaped), message);
  snprintf(json, sizeof(json), "{\"ok\":false,\"error\":\"%s\"}", escaped);
  respond_json(client, "422 Unprocessable Entity", NULL, json);
}

static void respond_user_session(int client, const char *email, const char *token) {
  char cookie[512];
  char escaped_email[512];
  char json[1024];
  json_escape(escaped_email, sizeof(escaped_email), email);
  snprintf(
    cookie,
    sizeof(cookie),
    "Set-Cookie: sealion_session=%s; HttpOnly; SameSite=Lax; Path=/; Max-Age=604800\r\n",
    token
  );
  snprintf(json, sizeof(json), "{\"ok\":true,\"user\":{\"email\":\"%s\"}}", escaped_email);
  respond_json(client, "200 OK", cookie, json);
}

void handle_register(int client, const Request *req) {
  char email[256];
  char password[256];
  char error[512];
  form_value(req->body, "email", email, sizeof(email));
  form_value(req->body, "password", password, sizeof(password));
  if (!create_user(email, password, error, sizeof(error))) {
    respond_auth_error(client, error);
    return;
  }
  int user_id = 0;
  lookup_user_id(email, &user_id);
  char token[129];
  if (!create_session(user_id, token, sizeof(token))) {
    respond_auth_error(client, "Account created, but login failed. Try logging in.");
    return;
  }
  respond_user_session(client, email, token);
}

void handle_login(int client, const Request *req) {
  char email[256];
  char password[256];
  int user_id = 0;
  form_value(req->body, "email", email, sizeof(email));
  form_value(req->body, "password", password, sizeof(password));
  if (!verify_user(email, password, &user_id)) {
    respond_auth_error(client, "Email or password is incorrect.");
    return;
  }
  char token[129];
  if (!create_session(user_id, token, sizeof(token))) {
    respond_auth_error(client, "Could not create a session.");
    return;
  }
  respond_user_session(client, email, token);
}

void handle_logout(int client, const Request *req) {
  destroy_session(req->cookie);
  respond_json(
    client,
    "200 OK",
    "Set-Cookie: sealion_session=deleted; HttpOnly; SameSite=Lax; Path=/; Max-Age=0\r\n",
    "{\"ok\":true}"
  );
}

File: templates/default/controller/auth_controller.go
package controller

import (
	"net/http"
	"time"

	"carbideapp/model"
)

const sessionCookie = "carbide_session"

func (a *App) handleRegister(w http.ResponseWriter, r *http.Request) {
	if !requireMethod(w, r, http.MethodPost) {
		return
	}
	if err := r.ParseForm(); err != nil {
		respondAuthError(w, "Could not read the form.")
		return
	}

	email := model.NormalizeEmail(r.FormValue("email"))
	password := r.FormValue("password")
	if err := a.Store.CreateUser(r.Context(), email, password); err != nil {
		respondAuthError(w, err.Error())
		return
	}

	userID, ok, err := a.Store.LookupUserID(r.Context(), email)
	if err != nil || !ok {
		respondAuthError(w, "Account created, but login failed. Try logging in.")
		return
	}
	token, err := a.Store.CreateSession(r.Context(), userID)
	if err != nil {
		respondAuthError(w, "Account created, but login failed. Try logging in.")
		return
	}
	respondUserSession(w, email, token)
}

func (a *App) handleLogin(w http.ResponseWriter, r *http.Request) {
	if !requireMethod(w, r, http.MethodPost) {
		return
	}
	if err := r.ParseForm(); err != nil {
		respondAuthError(w, "Could not read the form.")
		return
	}

	email := model.NormalizeEmail(r.FormValue("email"))
	userID, ok, err := a.Store.VerifyUser(r.Context(), email, r.FormValue("password"))
	if err != nil || !ok {
		respondAuthError(w, "Email or password is incorrect.")
		return
	}
	token, err := a.Store.CreateSession(r.Context(), userID)
	if err != nil {
		respondAuthError(w, "Could not create a session.")
		return
	}
	respondUserSession(w, email, token)
}

func (a *App) handleLogout(w http.ResponseWriter, r *http.Request) {
	if !requireMethod(w, r, http.MethodPost) {
		return
	}
	if cookie, err := r.Cookie(sessionCookie); err == nil {
		_ = a.Store.DestroySession(r.Context(), cookie.Value)
	}
	http.SetCookie(w, &http.Cookie{
		Name:     sessionCookie,
		Value:    "deleted",
		Path:     "/",
		MaxAge:   -1,
		HttpOnly: true,
		SameSite: http.SameSiteLaxMode,
	})
	respondJSON(w, http.StatusOK, map[string]any{"ok": true})
}

func respondAuthError(w http.ResponseWriter, message string) {
	respondJSON(w, http.StatusUnprocessableEntity, map[string]any{
		"ok":    false,
		"error": message,
	})
}

func respondUserSession(w http.ResponseWriter, email string, token string) {
	http.SetCookie(w, &http.Cookie{
		Name:     sessionCookie,
		Value:    token,
		Path:     "/",
		MaxAge:   int((7 * 24 * time.Hour).Seconds()),
		HttpOnly: true,
		SameSite: http.SameSiteLaxMode,
	})
	respondJSON(w, http.StatusOK, map[string]any{
		"ok":   true,
		"user": map[string]string{"email": email},
	})
}
